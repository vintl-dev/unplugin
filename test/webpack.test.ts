import {
  type Configuration as WebpackConfiguration,
  // eslint-disable-next-line import/no-named-default
  default as webpack,
  type StatsCompilation,
} from 'webpack'
import { describe, it, expect } from 'vitest'
import { createFsFromVolume, Volume } from 'memfs'
import type * as _distWebpack from '../dist/webpack'
import { createResolver } from './utils/resolver'

Error.stackTraceLimit = 1000

const { icuMessages }: typeof _distWebpack = await import(
  '../dist/webpa' + 'ck.cjs' // because unbuild emits .d.ts, but .(c|m)js files
)

const resolve = createResolver(import.meta.url)

class WebpackCompilationError extends Error {
  constructor(public causes?: unknown[]) {
    super('Compilation failed')
  }
}

async function buildFile(
  entry: string,
  extendConfig?: (config: WebpackConfiguration) => void | Promise<void>,
) {
  const config: WebpackConfiguration = {
    entry: resolve(entry),
    mode: 'production',
    output: {
      library: { type: 'module' },
      filename: 'index.mjs',
      path: '/dist',
    },
    experiments: { outputModule: true },
    optimization: { minimize: false },
  }

  await extendConfig?.(config)

  const compiler = webpack(config)

  const memfs = createFsFromVolume(new Volume())

  compiler.outputFileSystem = memfs

  await new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      let resolution: StatsCompilation | undefined
      let rejection: WebpackCompilationError | undefined

      if (err != null) {
        rejection = new WebpackCompilationError([err])
      } else if (stats != null) {
        const info: StatsCompilation = stats.toJson()

        const hasWarnings = stats.hasWarnings()
        const hasErrors = stats.hasErrors()

        if (hasErrors || hasWarnings) {
          if (hasWarnings) {
            console.warn('[webpack] [warn]', info.warnings)
          }

          rejection = new WebpackCompilationError(info.errors)
        } else {
          resolution = stats
        }
      }

      compiler.close((closeErr) => {
        if (closeErr != null) {
          if (rejection == null) {
            rejection = new WebpackCompilationError([closeErr])
          } else {
            ;(rejection.causes ??= []).push(closeErr)
          }
        }

        if (rejection != null) {
          reject(rejection)
          return
        }

        resolve(resolution!)
      })
    })
  })

  return memfs.readFileSync('/dist/index.mjs', { encoding: 'utf-8' })
}

describe(
  'webpack',
  () => {
    it('should generate bundle', async () => {
      const out = await buildFile('fixtures/normal/input.mjs', (config) => {
        ;(config.plugins ??= []).push(
          icuMessages({
            include: '**/*.messages.json',
            format: 'crowdin',
          }),
        )
      })

      expect(out).toMatchSnapshot()
    })

    it('should fail with unresolved formatter', async () => {
      let err: unknown
      try {
        await buildFile('fixtures/normal/input.mjs', (config) => {
          ;(config.plugins ??= []).push(
            icuMessages({
              include: '**/*.messages.json',
              format: 'non-existent',
            }),
          )
        })
      } catch (err_) {
        err = err_
      }

      expect(err).toBeDefined()
    })

    it('should be compatible with other json files', async () => {
      const out = await buildFile('fixtures/with-json/input.mjs', (config) => {
        ;(config.plugins ??= []).push(
          icuMessages({
            include: '**/*.messages.json',
            format: 'crowdin',
          }),
        )
      })

      expect(out).toMatchSnapshot()
    })

    it('should be able to parse using TOML files', async () => {
      const out = await buildFile('fixtures/toml/input.mjs', async (config) => {
        ;(config.plugins ??= []).push(
          icuMessages({
            include: '**/*.messages.toml',
            format: 'crowdin',
            parse: await import('@ltd/j-toml').then(
              (_) => (code: string) => _.parse(code),
            ),
          }),
        )
      })

      expect(out).toMatchSnapshot()
    })

    it('should transform to AST JSON', async () => {
      const out = await buildFile('fixtures/normal/input.mjs', (config) => {
        ;(config.plugins ??= []).push(
          icuMessages({
            include: '**/*.messages.json',
            format: 'crowdin',
            output: {
              format: 'json',
            },
          }),
        )
      })

      expect(out).toMatchSnapshot()
    })

    it('should transform to JSON of raw messages', async () => {
      const out = await buildFile('fixtures/normal/input.mjs', (config) => {
        ;(config.plugins ??= []).push(
          icuMessages({
            include: '**/*.messages.json',
            format: 'crowdin',
            output: {
              type: 'raw',
              format: 'json',
            },
          }),
        )
      })

      expect(out).toMatchSnapshot()
    })
  },
  { timeout: 60_000 },
)
