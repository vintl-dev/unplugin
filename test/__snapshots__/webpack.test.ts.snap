// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webpack > should be able to parse using TOML files 1`] = `
"/******/ // The require scope
/******/ var __webpack_require__ = {};
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/ 
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  \\"Z\\": () => (/* binding */ example)
});

;// CONCATENATED MODULE: ./test/fixtures/toml/en.messages.toml
const greeting = [
	{
		type: 0,
		value: \\"Hello, \\"
	},
	{
		type: 1,
		value: \\"name\\"
	},
	{
		type: 0,
		value: \\"!\\"
	}
];
/* harmony default export */ const en_messages = ({
	greeting: greeting
});

;// CONCATENATED MODULE: ./test/fixtures/toml/input.mjs


function example() {
  return en_messages.greeting
}

var __webpack_exports__default = __webpack_exports__.Z;
export { __webpack_exports__default as default };
"
`;

exports[`webpack > should be compatible with other json files 1`] = `
"/******/ // The require scope
/******/ var __webpack_require__ = {};
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/ 
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  \\"Z\\": () => (/* binding */ example)
});

;// CONCATENATED MODULE: ./test/fixtures/with-json/options.json
const options_namespaceObject = {\\"d\\":0};
;// CONCATENATED MODULE: ./test/fixtures/with-json/en.messages.json
const greeting = [
	{
		type: 0,
		value: \\"Hello, \\"
	},
	{
		type: 1,
		value: \\"name\\"
	},
	{
		type: 0,
		value: \\"!\\"
	}
];
const goodbye = [
	{
		type: 0,
		value: \\"Farewell, \\"
	},
	{
		type: 1,
		value: \\"name\\"
	},
	{
		type: 0,
		value: \\"!\\"
	}
];
/* harmony default export */ const en_messages = ({
	greeting: greeting,
	goodbye: goodbye
});

;// CONCATENATED MODULE: ./test/fixtures/with-json/input.mjs



const messageTypes = [
  [0, 'greeting'],
  [1, 'goodbye'],
]

function example() {
  const messageId = messageTypes.find(([type]) => options_namespaceObject.d === type)
  if (messageId == null) throw new Error(\`Unknown type \${options_namespaceObject.d}\`)
  return en_messages[messageId]
}

var __webpack_exports__default = __webpack_exports__.Z;
export { __webpack_exports__default as default };
"
`;

exports[`webpack > should generate bundle 1`] = `
"/******/ // The require scope
/******/ var __webpack_require__ = {};
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/ 
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  \\"Z\\": () => (/* binding */ example)
});

;// CONCATENATED MODULE: ./test/fixtures/normal/en.messages.json
const greeting = [
	{
		type: 0,
		value: \\"Hello, \\"
	},
	{
		type: 1,
		value: \\"name\\"
	},
	{
		type: 0,
		value: \\"!\\"
	}
];
/* harmony default export */ const en_messages = ({
	greeting: greeting
});

;// CONCATENATED MODULE: ./test/fixtures/normal/input.mjs


function example() {
  return en_messages
}

var __webpack_exports__default = __webpack_exports__.Z;
export { __webpack_exports__default as default };
"
`;

exports[`webpack > should transform to AST JSON 1`] = `
"/******/ // The require scope
/******/ var __webpack_require__ = {};
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/ 
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  \\"Z\\": () => (/* binding */ example)
});

;// CONCATENATED MODULE: ./test/fixtures/normal/en.messages.json
const en_messages_namespaceObject = JSON.parse('{\\"greeting\\":[{\\"type\\":0,\\"value\\":\\"Hello, \\"},{\\"type\\":1,\\"value\\":\\"name\\"},{\\"type\\":0,\\"value\\":\\"!\\"}]}');
;// CONCATENATED MODULE: ./test/fixtures/normal/input.mjs


function example() {
  return en_messages_namespaceObject
}

var __webpack_exports__default = __webpack_exports__.Z;
export { __webpack_exports__default as default };
"
`;

exports[`webpack > should transform to JSON of raw messages 1`] = `
"/******/ // The require scope
/******/ var __webpack_require__ = {};
/******/ 
/************************************************************************/
/******/ /* webpack/runtime/define property getters */
/******/ (() => {
/******/ 	// define getter functions for harmony exports
/******/ 	__webpack_require__.d = (exports, definition) => {
/******/ 		for(var key in definition) {
/******/ 			if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 				Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 			}
/******/ 		}
/******/ 	};
/******/ })();
/******/ 
/******/ /* webpack/runtime/hasOwnProperty shorthand */
/******/ (() => {
/******/ 	__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
/******/ })();
/******/ 
/************************************************************************/
var __webpack_exports__ = {};

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  \\"Z\\": () => (/* binding */ example)
});

;// CONCATENATED MODULE: ./test/fixtures/normal/en.messages.json
const en_messages_namespaceObject = JSON.parse('{\\"greeting\\":\\"Hello, {name}!\\"}');
;// CONCATENATED MODULE: ./test/fixtures/normal/input.mjs


function example() {
  return en_messages_namespaceObject
}

var __webpack_exports__default = __webpack_exports__.Z;
export { __webpack_exports__default as default };
"
`;
